src/Sulu/Bundle/SecurityBundle/Command/SyncPhpcrPermissionsCommand.php

```diff
@@ -0,0 +1,147 @@
+
+    /**
+     * @var DoctrineAccessControlProvider
+     */
+    private $doctrineAccessControlProvider;
+
+    public function __construct(
+        EntityManagerInterface $entityManager,
+        DocumentManagerInterface $documentManager,
+        DoctrineAccessControlProvider $doctrineAccessControlProvider
+    ) {
+        $this->entityManager = $entityManager;
+        $this->documentManager = $documentManager;
+        $this->doctrineAccessControlProvider = $doctrineAccessControlProvider;
+
+        parent::__construct();
+    }
+
+    public function configure(): void
+    {
+        $this->setDescription('Sync object permissions from phpcr into database');
+        $this->setHelp(
+            'The <info>%command.name%</info> command syncs the object permissions of phpcr documents into the database.'
+        );
+    }
+
+    public function execute(InputInterface $input, OutputInterface $output): int
+    {
+        $io = new SymfonyStyle($input, $output);
+
+        $roleIds = $this->getRoleIds();
+
+        if (0 === \count($roleIds)) {
+            $io->success('Nothing to do');
+
+            return 0;
+        }
+
+        $documents = $this->findDocuments($roleIds);
+        $amountDocuments = $documents->count();
+
+        if (0 === $amountDocuments) {
+            $io->success('Nothing to do');
+
+            return 0;
+        }
+
+        $io->progressStart($amountDocuments);
+
+        $count = 0;
+        foreach ($documents as $document) {
+            $io->progressAdvance();
+
+            if (!$document instanceof SecurityBehavior || !$document instanceof UuidBehavior) {
+                continue;
+            }
+
+            $this->doctrineAccessControlProvider->setPermissions(
+                SecurityBehavior::class,
+                $document->getUuid(),
+                $document->getPermissions()
+            );
+
+            ++$count;
+        }
+
+        $io->progressFinish();
+
+        $io->success(\sprintf('Successfully synced %d documents', $count));
+
+        return 0;
+    }
+
+    /**
+     * @return int[]
+     */
+    private function getRoleIds(): array
+    {
+        $result = $this->entityManager->createQueryBuilder()
+            ->select('role.id')
+            ->from(RoleInterface::class, 'role')
+            ->getQuery()
+            ->getArrayResult();
+
+        return \array_column($result, 'id');
+    }
+
+    /**
+     * Find documents with permissions for at least one of the role ids.
+     *
+     * @param int[] $roleIds
+     *
+     * @return QueryResultCollection<object>
+     */
+    protected function findDocuments(array $roleIds): QueryResultCollection
+    {
+        $roleIdsConstraint = \implode(' OR ', \array_map(function($roleId) {
+            return \sprintf('[%s%s] IS NOT NULL', SecuritySubscriber::SECURITY_PROPERTY_PREFIX, $roleId);
+        }, $roleIds));
+
+        $sql2 = \sprintf(
+            'SELECT * FROM [nt:unstructured] AS a WHERE (%s)',
```

Comment:
Should we do here some batching to avoid problems if a lot of nodes are loaded here?

Reply:
I don't think this is necessary, because the `QueryResultCollection` hydrates the documents one by one in the `current()` method, not all at the beginning

Reply:
Sounds good and as we are only selecting pages which actually have permission this should not be a too big problem